File: engines/unity/flutter_plugin/ios/cytoid_game_core/Sources/cytoid_game_core/MockGameCoreBridge.swift
import Foundation

final class MockGameCoreBridge {
  private static let protocolSchemaV2 = "cytoid.game-core.v2"
  private let emit: (String) -> Void
  private let hostReadyDelay: TimeInterval = 0.3
  private let resultDelay: TimeInterval = 1.0

  // Protocol smoke fake: it never emits session.telemetry; results always carry
  // a false/0/0 telemetry summary, including sessions with auto-class mods.
  private let runtimeState = RuntimeStateMachine()
  private var pendingResultWorkItem: DispatchWorkItem?

  init(emit: @escaping (String) -> Void) {
    self.emit = emit
  }

  func ensureRuntimeStarted() {
    let isFirstStart = runtimeState.state == .unavailable
    guard isFirstStart || runtimeState.state == .starting else { return }
    runtimeState.onRequestStart()
    if isFirstStart {
      DispatchQueue.main.asyncAfter(deadline: .now() + hostReadyDelay) { [weak self] in self?.emitHostReady() }
    }
  }

  func showGameSurface() {
    if runtimeState.state == .suspended { runtimeState.onResume() }
    ensureRuntimeStarted()
  }

  func hideGameSurface() {
    runtimeState.onSuspend()
  }

  func onOutboundMessage(_ jsonString: String) {
    guard let data = jsonString.data(using: .utf8),
      let envelope = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      envelope["schema"] as? String == Self.protocolSchemaV2,
      let type = envelope["type"] as? String else { return }
    switch type {
    case "session.start": handleSessionStart(envelope)
    case "session.cancel": handleSessionCancel(envelope)
    case "health.check": handleHealthCheck(envelope)
    case "settings.apply": handleSettingsApply(envelope)
    default: break
    }
  }

  private func handleSessionStart(_ envelope: [String: Any]) {
    guard let sessionId = envelope["id"] as? String else { return }
    guard let launchPayload = envelope["payload"] as? [String: Any] else {
      emitSessionResult(sessionId: sessionId, payload: rejectedResultPayload(
        sessionId: sessionId, mode: "unknown", mods: [], usedAutoMod: false,
        message: "session.start payload must be an object"))
      return
    }

    let mode = launchPayload["mode"] as? String
    let mods = stringArray(launchPayload["mods"])
    let usedAutoMod = containsAutoMod(mods)
    guard let mode, isSupportedMode(mode) else {
      emitSessionResult(sessionId: sessionId, payload: rejectedResultPayload(
        sessionId: sessionId, mode: mode ?? "unknown", mods: mods, usedAutoMod: usedAutoMod,
        message: "session.start mode is missing or unsupported"))
      return
    }

    runtimeState.onSessionStarted(sessionId: sessionId)
    emitSampleLogs(sessionId: sessionId)
    emit(["schema": Self.protocolSchemaV2, "id": sessionId, "type": "session.started", "payload": [
      "sessionId": sessionId, "mode": mode, "generation": runtimeState.generation,
    ]])

    let resultPayload = buildDefaultResult(
      sessionId: sessionId, mode: mode, mods: mods, usedAutoMod: usedAutoMod, launchPayload: launchPayload)
    let workItem = DispatchWorkItem { [weak self] in
      guard let self else { return }
      guard self.runtimeState.activeSessionId == sessionId else { return }
      self.runtimeState.onSessionEnded()
      self.emitSessionResult(sessionId: sessionId, payload: resultPayload)
    }
    pendingResultWorkItem = workItem
    DispatchQueue.main.asyncAfter(deadline: .now() + resultDelay, execute: workItem)
  }

  private func buildDefaultResult(
    sessionId: String, mode: String, mods: [String], usedAutoMod: Bool, launchPayload: [String: Any]
  ) -> [String: Any] {
    var payload = resultBase(
      sessionId: sessionId, mode: mode, mods: mods,
      outcome: ["kind": isCalibrationMode(mode) ? "calibration" : "completed"], usedAutoMod: usedAutoMod)
    if isCalibrationMode(mode) {
      payload["calibration"] = ["baseNoteOffset": 0.0, "levelNoteOffset": 0.0]
      return payload
    }
    payload["level"] = (launchPayload["level"] as? [String: Any]) ?? [
      "id": "mock-level", "title": "Mock Level", "difficulty": "mock", "difficultyLevel": 1,
    ]
    if mode == "tier" {
      let tierResult = buildMockTierResult(tier: launchPayload["tier"] as? [String: Any] ?? [:])
      payload["tier"] = tierResult.tier
      payload["score"] = scorePayload(maxCombo: tierResult.combo)
    } else {
      payload["score"] = scorePayload(maxCombo: 1_234)
    }
    return payload
  }

  private func buildMockTierResult(tier: [String: Any]) -> (tier: [String: Any], combo: Int) {
    let maxHealth = (tier["maxHealth"] as? NSNumber)?.doubleValue ?? 1_000
    let initialHealth = (tier["initialHealth"] as? NSNumber)?.doubleValue ?? maxHealth
    let combo = ((tier["initialCombo"] as? NSNumber)?.intValue ?? 0) + 50
    return (tier: [
      "tierId": tier["tierId"] as? String ?? "mock-tier",
      "stageIndex": (tier["stageIndex"] as? NSNumber)?.intValue ?? 0,
      "stageCount": (tier["stageCount"] as? NSNumber)?.intValue ?? 1,
      "health": max(initialHealth * 0.85, 0),
      "maxHealth": maxHealth,
      "combo": combo,
    ], combo: combo)
  }

  private func handleSessionCancel(_ envelope: [String: Any]) {
    guard let sessionId = envelope["id"] as? String else { return }
    guard runtimeState.activeSessionId == sessionId else { return }
    pendingResultWorkItem?.cancel()
    pendingResultWorkItem = nil
    let payload = envelope["payload"] as? [String: Any] ?? [:]
    let mode = payload["mode"] as? String ?? "ranked"
    let mods = stringArray(payload["mods"])
    runtimeState.onSessionEnded()
    emitSessionResult(sessionId: sessionId, payload: resultBase(
      sessionId: sessionId, mode: mode, mods: mods,
      outcome: ["kind": "cancelled", "reason": payload["reason"] as? String ?? "unknown"],
      usedAutoMod: containsAutoMod(mods)))
  }

  private func handleHealthCheck(_ envelope: [String: Any]) {
    guard let id = envelope["id"] as? String else { return }
    var payload: [String: Any] = [
      "engine": "mock", "generation": runtimeState.generation, "state": runtimeState.state.wireName,
    ]
    if let activeSessionId = runtimeState.activeSessionId { payload["activeSessionId"] = activeSessionId }
    emit(["schema": Self.protocolSchemaV2, "id": id, "type": "health.ok", "payload": payload])
  }

  private func handleSettingsApply(_ envelope: [String: Any]) {
    guard let id = envelope["id"] as? String else { return }
    let payload = envelope["payload"] as? [String: Any] ?? [:]
    emit(["schema": Self.protocolSchemaV2, "id": id, "type": "settings.applied", "payload": [
      "applied": true,
      "appliedFields": payload.keys.sorted(),
      "deferredFields": [],
      "rejectedFields": [],
      "errors": [],
    ]])
  }

  private func emitSampleLogs(sessionId: String) {
    let samples: [(String, String, String?)] = [
      ("debug", "Mock runtime started session", nil),
      ("warning", "Mock Unity: storyboard texture cache miss", nil),
      ("error", "Mock Unity: Unity artifact not mounted", "MockGameCoreBridge.swift:handleSessionStart"),
    ]
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) { [weak self] in
      let now = Self.epochMilliseconds()
      let logs = samples.map { sample -> [String: Any] in
        var entry: [String: Any] = [
          "level": sample.0, "message": sample.1, "timestamp": now, "sessionId": sessionId,
        ]
        if let stackTrace = sample.2 { entry["stackTrace"] = stackTrace }
        return entry
      }
      self?.emit(["schema": Self.protocolSchemaV2, "id": UUID().uuidString, "type": "logs.batch", "payload": [
        "reason": "trigger", "triggerLevel": "error", "timestamp": now, "truncated": false, "logs": logs,
      ]])
    }
  }

  private func emitHostReady() {
    runtimeState.onEngineReady()
    emit(["schema": Self.protocolSchemaV2, "id": UUID().uuidString, "type": "engine.ready", "payload": [
      "engine": "mock",
      "engineVersion": "cytoid_game_core",
      "generation": runtimeState.generation,
      "display": ["targetFrameRate": 60, "screenRefreshRate": 60],
    ]])
  }

  private func emitSessionResult(sessionId: String, payload: [String: Any]) {
    emit(["schema": Self.protocolSchemaV2, "id": sessionId, "type": "session.result", "payload": payload])
  }

  private func rejectedResultPayload(
    sessionId: String, mode: String, mods: [String], usedAutoMod: Bool, message: String
  ) -> [String: Any] {
    var payload = resultBase(
      sessionId: sessionId, mode: mode, mods: mods, outcome: ["kind": "rejected"], usedAutoMod: usedAutoMod)
    payload["error"] = ["code": "invalid_payload", "message": message]
    return payload
  }

  private func resultBase(
    sessionId: String, mode: String, mods: [String], outcome: [String: Any], usedAutoMod: Bool
  ) -> [String: Any] {
    [
      "sessionId": sessionId,
      "mode": mode,
      "mods": mods,
      "outcome": outcome,
      "flags": ["usedAutoMod": usedAutoMod],
      "telemetry": ["available": false, "eventsRecorded": 0, "bytes": 0],
      "timestamp": Self.epochMilliseconds(),
    ]
  }

  private func scorePayload(maxCombo: Int) -> [String: Any] {
    [
      "score": 950_000,
      "accuracy": 0.97,
      "maxCombo": maxCombo,
      "gradeCounts": ["perfect": 1_000, "great": 20, "good": 0, "bad": 0, "miss": 0],
      "early": 0,
      "late": 0,
      "averageTimingError": 0.0,
      "standardTimingError": 0.0,
    ]
  }

  private func stringArray(_ value: Any?) -> [String] {
    (value as? [Any])?.compactMap { $0 as? String } ?? []
  }

  private func containsAutoMod(_ mods: [String]) -> Bool {
    let autoMods = Set(["auto", "autodrag", "autohold", "autoflick"])
    return mods.contains { autoMods.contains($0.replacingOccurrences(of: "_", with: "").lowercased()) }
  }

  private func isSupportedMode(_ mode: String) -> Bool {
    mode == "ranked" || mode == "practice" || mode == "tier" || isCalibrationMode(mode)
  }

  private func isCalibrationMode(_ mode: String) -> Bool {
    mode == "calibration" || mode == "globalCalibration"
  }

  private func emit(_ envelope: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: envelope),
      let json = String(data: data, encoding: .utf8) else { return }
    DispatchQueue.main.async { [emit] in emit(json) }
  }

  private static func epochMilliseconds() -> Int {
    Int(Date().timeIntervalSince1970 * 1_000)
  }
}

File: engines/unity/flutter_plugin/android/src/main/kotlin/org/cytoid/gamecore/MockGameCoreBridge.kt
package org.cytoid.gamecore

import android.os.Handler
import android.os.Looper
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

class MockGameCoreBridge(
    private val emit: (String) -> Unit,
    private val mainHandler: Handler = Handler(Looper.getMainLooper()),
) {
    private val runtimeState = RuntimeStateMachine()

    fun ensureRuntimeStarted() {
        runtimeState.onRequestStart()
        if (runtimeState.state != RuntimeState.STARTING && runtimeState.state != RuntimeState.READY) return
        if (runtimeState.state == RuntimeState.STARTING) mainHandler.postDelayed({ emitHostReady() }, HOST_READY_DELAY_MS)
    }

    fun showGameSurface() {
        if (runtimeState.state == RuntimeState.SUSPENDED) runtimeState.onResume()
        ensureRuntimeStarted()
    }

    fun hideGameSurface() = runtimeState.onSuspend()

    fun onOutboundMessage(jsonString: String) {
        try {
            val envelope = JSONObject(jsonString)
            when (envelope.getString("type")) {
                "session.start" -> handleSessionStart(envelope)
                "session.cancel" -> handleSessionCancel(envelope)
                "health.check" -> handleHealthCheck(envelope)
                "settings.apply" -> handleSettingsApply(envelope)
                else -> Log.w(TAG, "Unhandled outbound type: ${envelope.getString("type")}")
            }
        } catch (error: Exception) {
            Log.e(TAG, "Failed to parse outbound envelope", error)
        }
    }

    private fun handleSessionStart(envelope: JSONObject) {
        val sessionId = envelope.getString("id")
        val launch = envelope.optJSONObject("payload")
        val mode = launch?.optString("mode").orEmpty()
        val mods = launch?.optJSONArray("mods") ?: JSONArray()

        if (!isValidLaunch(launch, mode)) {
            val rejectedEnvelope = envelope(sessionId, "session.result", rejectedResult(sessionId, mode, mods)).toString()
            emit(rejectedEnvelope)
            return
        }

        val validLaunch = launch ?: return
        runtimeState.onSessionStarted(sessionId)
        emitEnvelope(envelope(sessionId, "session.started", sessionStarted(sessionId, mode)).toString())
        emitSampleLogs(sessionId)
        scheduleResult(sessionId, defaultResult(sessionId, mode, mods, validLaunch))
    }

    private fun isValidLaunch(launch: JSONObject?, mode: String): Boolean =
        launch != null && mode in SUPPORTED_MODES && launch.optJSONArray("mods") != null &&
            launch.optJSONObject("options") != null && (mode != "tier" || launch.optJSONObject("tier") != null)

    private fun sessionStarted(sessionId: String, mode: String): JSONObject =
        obj("sessionId" to sessionId, "mode" to mode, "generation" to runtimeState.generation)

    private fun scheduleResult(sessionId: String, payload: JSONObject) {
        mainHandler.postDelayed({
            if (runtimeState.activeSessionId != sessionId) return@postDelayed
            runtimeState.onSessionEnded()
            emitEnvelope(envelope(sessionId, "session.result", payload).toString())
        }, MOCK_GAME_RESULT_DELAY_MS)
    }

    private fun defaultResult(
        sessionId: String,
        mode: String,
        mods: JSONArray,
        launch: JSONObject,
    ): JSONObject {
        val payload = baseResult(sessionId, mode, mods).put("outcome", obj("kind" to defaultOutcome(mode)))
        when (mode) {
            "ranked", "practice" -> payload.put("level", levelResult(launch)).put("score", scoreResult())
            "tier" -> payload
                .put("level", levelResult(launch))
                .put("score", scoreResult())
                .put("tier", tierResult(launch.getJSONObject("tier")))
            "calibration", "globalCalibration" -> payload.put(
                "calibration",
                obj("baseNoteOffset" to 0.0, "levelNoteOffset" to 0.0),
            )
        }
        return payload
    }

    private fun defaultOutcome(mode: String): String = if (mode == "calibration" || mode == "globalCalibration") "calibration" else "completed"

    private fun baseResult(
        sessionId: String,
        mode: String,
        mods: JSONArray,
    ): JSONObject =
        obj(
            "sessionId" to sessionId,
            "mode" to mode.ifEmpty { "ranked" },
            "mods" to copyArray(mods),
            "flags" to obj("usedAutoMod" to containsAutoClassMod(mods)),
            // The smoke mock records no play events and emits no session.telemetry envelope.
            "telemetry" to obj("available" to false, "eventsRecorded" to 0, "bytes" to 0),
            "timestamp" to System.currentTimeMillis(),
        )

    private fun levelResult(launch: JSONObject): JSONObject {
        val level = launch.optJSONObject("level") ?: return obj("id" to "mock-level")
        val meta = level.optJSONObject("meta")
        val selected = level.optString("selectedDifficulty", "hard")
        val chart = selectedChart(meta, selected)
        return obj(
            "id" to (meta?.optString("id", "mock-level") ?: "mock-level"),
            "title" to (meta?.optString("title", "Mock Level") ?: "Mock Level"),
            "difficulty" to selected,
            "difficultyLevel" to (chart?.optInt("difficulty", 1) ?: 1),
        )
    }

    private fun selectedChart(meta: JSONObject?, selected: String): JSONObject? {
        val charts = meta?.optJSONArray("charts") ?: return null
        for (index in 0 until charts.length()) {
            val chart = charts.optJSONObject(index) ?: continue
            if (chart.optString("type") == selected) return chart
        }
        return null
    }

    private fun scoreResult(): JSONObject =
        obj(
            "score" to 950000,
            "accuracy" to 0.97,
            "maxCombo" to 50,
            "gradeCounts" to obj("perfect" to 900, "great" to 80, "good" to 15, "bad" to 3, "miss" to 2),
            "early" to 4,
            "late" to 6,
            "averageTimingError" to 0.0,
            "standardTimingError" to 12.5,
        )

    private fun tierResult(tier: JSONObject): JSONObject {
        val maxHealth = tier.optDouble("maxHealth", 1000.0)
        val health = tier.optDouble("health", tier.optDouble("initialHealth", maxHealth))
        val combo = tier.optInt("combo", tier.optInt("initialCombo", 0))
        return obj(
            "tierId" to tier.optString("tierId", "mock-tier"),
            "stageIndex" to tier.optInt("stageIndex", 0),
            "stageCount" to tier.optInt("stageCount", 1),
            "health" to (health * 0.85).coerceAtLeast(0.0),
            "maxHealth" to maxHealth,
            "combo" to combo + 50,
        )
    }

    private fun rejectedResult(
        sessionId: String,
        mode: String,
        mods: JSONArray,
    ): JSONObject =
        baseResult(sessionId, mode, mods)
            .put("outcome", obj("kind" to "rejected"))
            .put("error", obj("code" to "invalid_payload", "message" to "Mock session.start payload is structurally invalid"))

    private fun handleSessionCancel(envelope: JSONObject) {
        val sessionId = envelope.getString("id")
        val reason = envelope.optJSONObject("payload")?.optString("reason", "unknown") ?: "unknown"
        runtimeState.onSessionEnded()
        emitEnvelope(
            envelope(
                sessionId,
                "session.result",
                baseResult(sessionId, "ranked", JSONArray()).put("outcome", obj("kind" to "cancelled", "reason" to reason)),
            ).toString(),
        )
    }

    private fun handleHealthCheck(envelope: JSONObject) {
        val payload = obj("engine" to "mock", "generation" to runtimeState.generation, "state" to runtimeState.state.wireName)
        runtimeState.activeSessionId?.let { payload.put("activeSessionId", it) }
        emitEnvelope(envelope(envelope.getString("id"), "health.ok", payload).toString())
    }

    private fun handleSettingsApply(envelope: JSONObject) {
        val appliedFields = JSONArray()
        (envelope.optJSONObject("payload") ?: JSONObject()).keys().forEach { appliedFields.put(it) }
        emitEnvelope(
            envelope(
                envelope.getString("id"),
                "settings.applied",
                obj(
                    "applied" to true,
                    "appliedFields" to appliedFields,
                    "deferredFields" to JSONArray(),
                    "rejectedFields" to JSONArray(),
                    "errors" to JSONArray(),
                ),
            ).toString(),
        )
    }

    private fun emitSampleLogs(sessionId: String) {
        val samples = listOf(
            Triple("log", "Mock runtime accepted session", null),
            Triple("warning", "Mock Unity: storyboard texture cache miss", null),
            Triple("error", "Mock Unity: Unity artifact not mounted", "MockGameCoreBridge.kt:handleSessionStart"),
        )
        mainHandler.postDelayed({
            val logs = JSONArray()
            samples.forEach { (level, message, stackTrace) ->
                val entry = obj("level" to logLevel(level), "message" to message, "timestamp" to System.currentTimeMillis(), "sessionId" to sessionId)
                stackTrace?.let { entry.put("stackTrace", it) }
                logs.put(entry)
            }
            emitEnvelope(
                envelope(
                    UUID.randomUUID().toString(),
                    "logs.batch",
                    obj("reason" to "trigger", "triggerLevel" to "error", "timestamp" to System.currentTimeMillis(), "truncated" to false, "logs" to logs),
                ).toString(),
            )
        }, MOCK_LOG_DELAY_MS)
    }

    private fun logLevel(level: String): String =
        when (level) {
            "log" -> "debug"
            "warning" -> "warning"
            "error", "exception" -> "error"
            else -> level
        }

    private fun emitHostReady() {
        runtimeState.onEngineReady()
        emitEnvelope(
            envelope(
                UUID.randomUUID().toString(),
                "engine.ready",
                obj(
                    "engine" to "mock",
                    "engineVersion" to "cytoid_game_core",
                    "generation" to runtimeState.generation,
                    "display" to obj("targetFrameRate" to 60, "screenRefreshRate" to 60),
                ),
            ).toString(),
        )
    }

    private fun envelope(id: String, type: String, payload: JSONObject): JSONObject = obj("schema" to PROTOCOL_SCHEMA, "id" to id, "type" to type, "payload" to payload)

    private fun containsAutoClassMod(mods: JSONArray): Boolean {
        for (index in 0 until mods.length()) if (mods.optString(index) in AUTO_CLASS_MODS) return true
        return false
    }

    private fun copyArray(source: JSONArray): JSONArray {
        val copy = JSONArray()
        for (index in 0 until source.length()) copy.put(source.opt(index))
        return copy
    }

    private fun obj(vararg pairs: Pair<String, Any?>): JSONObject = JSONObject().apply { pairs.forEach { put(it.first, it.second) } }

    private fun emitEnvelope(json: String) = emit(json)

    companion object {
        private const val TAG = "MockGameCoreBridge"
        private const val PROTOCOL_SCHEMA = "cytoid.game-core.v2"
        private const val HOST_READY_DELAY_MS = 150L
        private const val MOCK_GAME_RESULT_DELAY_MS = 900L
        private const val MOCK_LOG_DELAY_MS = 600L
        private val SUPPORTED_MODES = setOf("ranked", "practice", "calibration", "globalCalibration", "tier")
        private val AUTO_CLASS_MODS = setOf("auto", "autoDrag", "autoHold", "autoFlick")
    }
}
